historical_data_processing/archive_processing_2013-2015.py

```diff
@@ -0,0 +1,229 @@
+    "datahub.austintexas.gov",
+    SOCRATA_APP_TOKEN,
+    username=SOCRATA_API_KEY_ID,
+    password=SOCRATA_API_KEY_SECRET,
+    timeout=60,
+)
+
+dbx = dropbox.Dropbox(METROBIKE_DROPBOX_TOKEN)
+
+
+def get_data(path, token):
+    """Get trip data file as string from dropbox"""
+    dbx = dropbox.Dropbox(token)
+
+    try:
+        metadata, res = dbx.files_download(path)
+
+    except dropbox.exceptions.ApiError:
+        #  file not found - we take that to mean the data has not been uploaded to Dropbox yet by Metrobike staff
+        # logger.warning(f"No trip data file found at {path}")
+        return None
+    res.raise_for_status()
+    return res
+
+
+# Downloading CSV from dropbox
+data = get_data(
+    "/austinbcycletripdata/AustinTripQuery_pre2018.csv", METROBIKE_DROPBOX_TOKEN
+)
+df = pd.read_csv(StringIO(data.text))
+
+# Mapping for columns that are already in the right format
+column_mapping = {
+    "TripId": "trip_id",
+    "MembershipType": "membership_type",
+    "BikeVisibleId": "bicycle_id",
+    "Duration": "trip_duration_minutes",
+}
+df.rename(columns=column_mapping, inplace=True)
+
+# Filtering out maintenance trips and short trips
+df = df[df["trip_duration_minutes"] > 1]
+df = df[df["ActualCharge"] >= 0]
+
+# Date/time fields
+central_timezone = timezone("US/Central")
+
+df["RentalDate CST"] = pd.to_datetime(df["RentalDate CST"].str[:-7])
+df["RentalDate CST"] = (
+    df["RentalDate CST"].dt.tz_localize("utc").dt.tz_convert(central_timezone)
+)
+df["checkout_date"] = df["RentalDate CST"].dt.strftime("%Y-%m-%d")
+df["checkout_time"] = df["RentalDate CST"].dt.strftime("%-H:%M:%S")
+df["checkout_datetime"] = df["RentalDate CST"].dt.strftime("%Y-%m-%dT%H:%M:%S")
+df["year"] = df["RentalDate CST"].dt.strftime("%Y")
+df["month"] = df["RentalDate CST"].dt.strftime("%-m")
+df["hour"] = df["RentalDate CST"].dt.strftime("%H")
+
+# Kiosk IDs lookup city asset number value in another table
+kiosks = pd.read_csv("https://data.austintexas.gov/resource/qd73-bsdg.csv")
+
+# Adding in some missing kiosk IDs into the dataset
+kiosks.loc[kiosks["kiosk_id"] == 2550, "city_asset_num"] = 2550
+kiosks.loc[kiosks["kiosk_id"] == 2576, "city_asset_num"] = 2576
+kiosks.loc[kiosks["kiosk_id"] == 2564, "city_asset_num"] = 2564
+kiosks.loc[kiosks["kiosk_id"] == 2536, "city_asset_num"] = 2536
+kiosks.loc[kiosks["kiosk_id"] == 3381, "city_asset_num"] = 3381
+kiosks.loc[kiosks["kiosk_id"] == 2546, "city_asset_num"] = 2546
+kiosks.loc[kiosks["kiosk_id"] == 2545, "city_asset_num"] = 2545
+kiosks.loc[kiosks["kiosk_id"] == 2712, "city_asset_num"] = 6
+kiosks.loc[kiosks["kiosk_id"] == 1006, "city_asset_num"] = 1006
+kiosks.loc[kiosks["kiosk_id"] == 2500, "city_asset_num"] = 2500
+
+# Create lookup table between city_asset_num used by the CSV and kiosk_ids we use in the dataset
+kiosks = kiosks[kiosks["city_asset_num"] >= 0]
+kiosks["city_asset_num"] = kiosks["city_asset_num"].astype("Int32")
+kiosks = kiosks[["kiosk_id", "city_asset_num"]]
```

Comment:
🐼 💪 - how this is rolling up these loops to slice and dice the data is really concise and readable. Really nice - TIL.